File: 103.binary-tree-zigzag-level-order-traversal.java
import java.util.LinkedList;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=103 lang=java
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> results = new ArrayList<List<Integer>>();
        if (root == null) {
            return results;
        }

        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);

        boolean reverse = false;
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> list = new ArrayList<Integer>();

            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (reverse) {
                    list.add(0, node.val);
                } else {
                    list.add(node.val);
                }

                // every time we get the node from the queue
                // we use a fixed length size to get the node 
                // from the head of the queue, add push the new node
                // to the tail of the queue
                if (node.left != null) {
                    queue.offer(node.left);
                }

                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            results.add(list);
            reverse = !reverse;
        }

        return results;
    }
}
// @lc code=end

File: 103.binary-tree-zigzag-level-order-traversal.cpp
/*
 * @lc app=leetcode id=103 lang=cpp
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) {
            return res;
        }

        queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);

        bool inverse = false;
        while (!nodeQueue.empty()) {
            int size = nodeQueue.size();
            vector<int> level;

            for (int i = 0; i < size; i++) {
                TreeNode* node = nodeQueue.front(); nodeQueue.pop();
                if (inverse) {
                    level.insert(level.begin(), node->val);
                } else {
                    level.push_back(node->val);
                }

                if (node->left) {
                    nodeQueue.push(node->left);        
                }

                if (node->right) {
                    nodeQueue.push(node->right);
                }
            }

            inverse = !inverse;
            res.push_back(level);
        }

        return res;
    }
};
// @lc code=end

